### ISSUE TITLE
Alternative for java-home not installed: failed to read link /usr/java/default: No such file or directory

### ISSUE THREAD

**Issue by porunov** [saltstack-formulas/sun-java-formula#49]
I am trying to run this formula on a fresh CentOS 7.3 Minimal. I am kind of confused about `/usr/java/default`. The problem is that my pillar file is identical to pillar file in this formula.

Pillar data (I.e. `salt '192.168.0.106' pillar.data`) shows that:
```
java_home:
        /usr/lib/java
```

Why does it try to read link `/usr/java/default`?

Here is the full output:

```
192.168.0.106:
----------
          ID: java-install-dir
    Function: file.directory
        Name: /usr/share/java
      Result: True
     Comment: Directory /usr/share/java is in the correct state
              Directory /usr/share/java updated
     Started: 01:34:39.409504
    Duration: 0.679 ms
     Changes:   
----------
          ID: sun-java-remove-prev-archive
    Function: file.absent
        Name: /usr/share/java/jdk-8u144-linux-x64.tar.gz
      Result: True
     Comment: File /usr/share/java/jdk-8u144-linux-x64.tar.gz is not present
     Started: 01:34:39.410503
    Duration: 0.285 ms
     Changes:   
----------
          ID: download-jdk-archive
    Function: cmd.run
        Name: curl -b oraclelicense=accept-securebackup-cookie -L -s -o '/usr/share/java/jdk-8u144-linux-x64.tar.gz' 'http://download.oracle.com/otn-pub/java/jdk/8u144-b01/090f390dda5b47b9b721c7dfaa008135/jdk-8u144-linux-x64.tar.gz'
      Result: True
     Comment: unless execution succeeded
     Started: 01:34:39.411421
    Duration: 8.361 ms
     Changes:   
----------
          ID: check-jdk-archive
    Function: module.run
        Name: file.check_hash
      Result: True
     Comment: State was not run because none of the onchanges reqs changed
     Changes:   
----------
          ID: unpack-jdk-archive
    Function: archive.extracted
        Name: /usr/share/java
      Result: True
     Comment: State was not run because none of the onchanges reqs changed
     Changes:   
----------
          ID: update-javahome-symlink
    Function: file.symlink
        Name: /usr/lib/java
      Result: True
     Comment: Symlink /usr/lib/java is present and owned by root:root
     Started: 01:34:39.421666
    Duration: 21.213 ms
     Changes:   
----------
          ID: remove-jdk-archive
    Function: file.absent
        Name: /usr/share/java/jdk-8u144-linux-x64.tar.gz
      Result: True
     Comment: File /usr/share/java/jdk-8u144-linux-x64.tar.gz is not present
     Started: 01:34:39.443101
    Duration: 0.337 ms
     Changes:   
----------
          ID: sun-java-jce-unzip
    Function: pkg.installed
        Name: unzip
      Result: True
     Comment: All specified packages are already installed
     Started: 01:34:39.848616
    Duration: 356.824 ms
     Changes:   
----------
          ID: sun-java-remove-old-jce-archive
    Function: file.absent
        Name: /usr/share/java/jdk1.8.0_144/jre/lib/security/UnlimitedJCEPolicy.zip
      Result: True
     Comment: File /usr/share/java/jdk1.8.0_144/jre/lib/security/UnlimitedJCEPolicy.zip is not present
     Started: 01:34:40.205840
    Duration: 0.333 ms
     Changes:   
----------
          ID: download-jce-archive
    Function: cmd.run
        Name: curl -b oraclelicense=accept-securebackup-cookie -L -s -o '/usr/share/java/jdk1.8.0_144/jre/lib/security/UnlimitedJCEPolicy.zip' 'http://download.oracle.com/otn-pub/java/jce/8/jce_policy-8.zip'
      Result: True
     Comment: onlyif execution failed
     Started: 01:34:40.206478
    Duration: 4.739 ms
     Changes:   
----------
          ID: check-jce-archive
    Function: module.run
        Name: file.check_hash
      Result: True
     Comment: State was not run because none of the onchanges reqs changed
     Changes:   
----------
          ID: backup-non-jce-jar
    Function: cmd.run
        Name: mv US_export_policy.jar US_export_policy.jar.nonjce; mv local_policy.jar local_policy.jar.nonjce;
      Result: True
     Comment: /usr/share/java/jdk1.8.0_144/jre/lib/security/US_export_policy.jar.nonjce exists
     Started: 01:34:40.211862
    Duration: 0.357 ms
     Changes:   
----------
          ID: unpack-jce-archive
    Function: cmd.run
        Name: unzip -j -o /usr/share/java/jdk1.8.0_144/jre/lib/security/UnlimitedJCEPolicy.zip
      Result: True
     Comment: /usr/share/java/jdk1.8.0_144/jre/lib/security/US_export_policy.jar exists
     Started: 01:34:40.212744
    Duration: 0.294 ms
     Changes:   
----------
          ID: remove-jce-archive
    Function: file.absent
        Name: /usr/share/java/jdk1.8.0_144/jre/lib/security/UnlimitedJCEPolicy.zip
      Result: True
     Comment: File /usr/share/java/jdk1.8.0_144/jre/lib/security/UnlimitedJCEPolicy.zip is not present
     Started: 01:34:40.213391
    Duration: 0.286 ms
     Changes:   
----------
          ID: jdk-config
    Function: file.managed
        Name: /etc/profile.d/java.sh
      Result: True
     Comment: File /etc/profile.d/java.sh is in the correct state
     Started: 01:34:40.213749
    Duration: 13.337 ms
     Changes:   
----------
          ID: javahome-alt-install
    Function: alternatives.install
        Name: java-home
      Result: False
     Comment: Alternative for java-home not installed: failed to read link /usr/java/default: No such file or directory
              the primary link for java-home must be /usr/java/default
     Started: 01:34:40.227477
    Duration: 5.392 ms
     Changes:   
----------
          ID: javahome-alt-set
    Function: alternatives.set
        Name: java-home
      Result: False
     Comment: One or more requisite failed: sun-java.env.javahome-alt-install
     Changes:   
----------
          ID: java-alt-install
    Function: alternatives.install
        Name: java
      Result: False
     Comment: One or more requisite failed: sun-java.env.javahome-alt-set
     Changes:   
----------
          ID: java-alt-set
    Function: alternatives.set
        Name: java
      Result: False
     Comment: One or more requisite failed: sun-java.env.java-alt-install
     Changes:   
----------
          ID: javac-alt-install
    Function: alternatives.install
        Name: javac
      Result: False
     Comment: One or more requisite failed: sun-java.env.java-alt-set
     Changes:   
----------
          ID: javac-alt-set
    Function: alternatives.set
        Name: javac
      Result: False
     Comment: One or more requisite failed: sun-java.env.javac-alt-install
     Changes:   

Summary for 192.168.0.106
-------------
Succeeded: 15
Failed:     6
-------------
Total states run:     21
Total run time:  412.437 ms

```

Any ideas?

**Comment by noelmcloughlin** [comment]
I did not replicate on Centos 7.3.1611 (Core). 
```
          ID: javahome-alt-install
    Function: alternatives.install
        Name: java-home
      Result: True
     Comment: Alternative for java-home set to path /usr/share/java/jdk1.8.0_144 with priority 301800111
     Started: 14:22:42.265430
    Duration: 72.539 ms
     Changes: 
              ----------
              link:
                  /usr/lib/java
              name:
                  java-home
              path:
                  /usr/share/java/jdk1.8.0_144
              priority:
                  301800111
```
I'll investigate more this week (see: https://bugzilla.redhat.com/show_bug.cgi?id=1324714)

**Comment by porunov** [comment]
It seems that the problem is in saltstack itself (I am using version 2017.7.1). Here is the issue:
https://github.com/saltstack/salt/issues/38802 

We had a discussion with @sroegner and @vutny about alternatives in `zookeeper-formula` (https://github.com/saltstack-formulas/zookeeper-formula/pull/38).
We decided to remove them because of the next facts:
1) They are broken in saltstack. It can work on some systems and on some systems it can be broken (depends on OS and version of saltstack).
2) We (maybe not completely) don't know cases where you need to use both alternatives and saltstack.
I.e. if you are using saltstack then your deployment is automatic. To install a new version of Java you just use the formula for it. (I.e. You are not installing Java 7 with saltstack and then Java 8 manually and switch alternatives to Java 8.).

I do understand that theoretically someone can depend on this feature, but in other side, I do think that it would be better to drop alternatives support at least before they are stable in saltstack.
I would just use simple symbolic links instead of alternatives.

@noelmcloughlin @sroegner @vutny What are your thoughts?

If needed I can prepare fixes for this formula or somebody else can do it.

**Comment by sroegner** [comment]
@porunov I am perfectly fine with getting rid of the alternatives system wherever possible.

**Comment by noelmcloughlin** [comment]
hi @porunov   Please do not get rid of alternatives code via current #50 proposal.  Please instead update the formula to support but not force alternatives.  There are few issues to fix #46, #47, #49 although I cannot replicate your issue across various Linux (Suse, Cent7, Arch, Ubuntu).

See this [commit](https://github.com/saltstack-formulas/sqlplus-formula/pull/11/commits/7a3d58388aa249654594a87913aaba5c11c7c3be ) for alternative solution which should satisfy needs of everybody without regressionl

**Comment by porunov** [comment]
The issue is fixed by #50
